codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -331,22 +331,26 @@ private Http2Stream requireStream(int streamId) {
      */
     private final class FlowControlledData extends FlowControlledBase {
         private final CoalescingBufferQueue queue;
+        private int dataSize;
 
         public FlowControlledData(Http2Stream stream, ByteBuf buf, int padding, boolean endOfStream,
                                    ChannelPromise promise) {
             super(stream, padding, endOfStream, promise);
-            queue = new CoalescingBufferQueue(promise.channel());
+            queue = new CoalescingBufferQueue(promise.channel(), 8);
             queue.add(buf, promise);
+            dataSize = queue.readableBytes();
         }
 
         @Override
         public int size() {
-            return queue.readableBytes() + padding;
+            return dataSize + padding;
         }
 
         @Override
         public void error(ChannelHandlerContext ctx, Throwable cause) {
             queue.releaseAndFailAll(cause);
+            // Don't update dataSize because we need to ensure the size() method returns a consistent size even after
```

Comment:
Not necessarily. If the stream has frames queued in the remote flow controller and the stream is closed (e.g. received a RST_STREAM) then all queued frames will be cancelled and this will directly call `FlowControlled.error(...)`.